jmclient/jmclient/payjoin.py

```diff
@@ -0,0 +1,503 @@
+    d = agent.request(b"POST", manager.server.encode("utf-8"),
+        Headers({"Content-Type": ['text/plain']}),
+        bodyProducer=body)
+
+    d.addCallback(receive_payjoin_proposal_from_server, manager)
+    # note that the errback (here "noresponse") is *not* triggered
+    # by a server rejection (which is accompanied by a non-200
+    # status code returned), but by failure to communicate.
+    def noResponse(failure):
+        failure.trap(ResponseFailed)
+        log.error(failure.value.reasons[0].getTraceback())
+        reactor.stop()
+    d.addErrback(noResponse)
+    return (True, None)
+
+def fallback_nonpayjoin_broadcast(manager, err):
+    assert isinstance(manager, JMPayjoinManager)
+    log.warn("Payjoin did not succeed, falling back to non-payjoin payment.")
+    log.warn("Error message was: " + str(err))
+    original_tx = manager.initial_psbt.extract_transaction()
+    if not jm_single().bc_interface.pushtx(original_tx.serialize()):
+        log.error("Unable to broadcast original payment. The payment is NOT made.")
+    log.info("We paid without coinjoin. Transaction: ")
+    log.info(btc.hrt(original_tx))
+    reactor.stop()
+
+def receive_payjoin_proposal_from_server(response, manager):
+    assert isinstance(manager, JMPayjoinManager)
+
+    # if the response code is not 200 OK, we must assume payjoin
+    # attempt has failed, and revert to standard payment.
+    if int(response.code) != 200:
+        fallback_nonpayjoin_broadcast(manager, err=response.phrase)
+        return
+    # for debugging; will be removed in future:
+    log.debug("Response headers:")
+    log.debug(pformat(list(response.headers.getAllRawHeaders())))
+    # no attempt at chunking or handling incrementally is needed
+    # here. The body should be a byte string containing the
+    # new PSBT.
+    d = readBody(response)
+    d.addCallback(process_payjoin_proposal_from_server, manager)
+
+def process_payjoin_proposal_from_server(response_body, manager):
+    assert isinstance(manager, JMPayjoinManager)
+    try:
+        payjoin_proposal_psbt = \
+            btc.PartiallySignedTransaction.from_base64(response_body)
+    except Exception as e:
+        log.error("Payjoin tx from server could not be parsed: " + repr(e))
+        fallback_nonpayjoin_broadcast(manager, err="Server sent invalid psbt")
+        return
+
+    log.debug("Receiver sent us this PSBT: ")
+    log.debug(manager.wallet_service.hr_psbt(payjoin_proposal_psbt))
+    # we need to add back in our utxo information to the received psbt,
+    # since the servers remove it (not sure why?)
+    for i, inp in enumerate(payjoin_proposal_psbt.unsigned_tx.vin):
+        for j, inp2 in enumerate(manager.initial_psbt.unsigned_tx.vin):
+                    if (inp.prevout.hash, inp.prevout.n) == (
+                        inp2.prevout.hash, inp2.prevout.n):
+                        payjoin_proposal_psbt.inputs[i].utxo = \
+                            manager.initial_psbt.inputs[j].utxo
```

Comment:
Direct assignment to utxo in input also won't work after https://github.com/Simplexum/python-bitcointx/pull/35 is merged, because assignment assignment needs to set `_witness_utxo` field in case if `utxo` is set to `CTransaction` but the input is segwit. `set_utxo()` methods is provided, which also requires appropriate `unsigned_tx` to be supplied to it

Reply:
Maybe there's a way to do this more ergonomically while still allowing `CTransaction` to be retained for witness inputs. I thought that direct-assignment to `utxo` field would be a rare occurence.

Reply:
I found myself wanting (needing?) to do it here because after already constructing the PSBT once, I had to reconstruct input-by-input once the server return the new PSBT, which represents a different actual tx, but using the same inputs of *mine* (the client) as before.

Updated to use `set_utxo()` in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/536/commits/9933408ad70e717014ecbb24796d104f3263ac05